hkmc2/shared/src/main/scala/hkmc2/semantics/Resolver.scala

```diff
@@ -0,0 +1,766 @@
+  * be present. For example, by-name, lazy and implicit parameters all
+  * require the definition so that the compiler can perform the correct
+  * elaboration and resolution.
+  *
+  * ### Module Methods & Modulefulness Check
+  *
+  * In addition to (global or local) functions, module methods also
+  * always have the definition present at compilation-time. We will now
+  * start to refer to the module methods as "functions".
+  *
+  * The definition of module functions are always present because of
+  * some restrictions on modules. Only declarations that have a
+  * signature and marked as `module` may bind to a module value. For
+  * example, 
+  *
+  * ```mls
+  * val v: module M
+  * fun f(module m: M): module M
+  * ```
+  *
+  * This preserves the module definition even if the module is bound to
+  * other names. In order to maintain the correctness of this system,
+  * the resolver performs the extra checks to ensure that the definition
+  * of a module is never lost during re-binding. 
+  *
+  * 1. Only declarations that have a signature and marked as `module`
+  *    may bind to a module value.
+  * 2. Only declarations of static members may bind to a module value.
+  *
+  * If a term evaluates to a module value (e.g., an application to a
+  * function returning a module, a reference to a variable of module),
+  * it is said to be moduleful. All moduleful terms must only occur at
+  * certain locations in the program. For example, a moduleful term must
+  * not occur as a scrutinee because it may be re-bound to a different
+  * name, during when the module definition is lost.
+  *
+  * ### Symbol Resolution
+  *
+  * The resolver also resolves extra symbols that are not resolved
+  * during the elaboration. Consider the following example:
+  *
+  * ```mls
+  * val m: module M = M.v.v
+  * module M with
+  *   val v: module M = M
+  * ```
+  *
+  * At the time of elaboration on `M.v.v`, the elaborator hasn't
+  * elaborated the module `M` yet, so it merely knows that `M` is a
+  * module and `v` is a member of `M`. It doesn't know that `M.v` is a
+  * module, so it cannot resolve `M.v.v` to the correct member symbol.
+  * Therefore, resolution of such symbols is deferred to the resolver. 
+  */
+class Resolver(tl: TraceLogger)
+(using raise: Raise, state: State):
+  import tl.*
+  import Resolver.ICtx
+  import Resolver.ictx
+  
+  enum Expect:
+    case Module(reason: Opt[Message])
+    case NonModule(reason: Opt[Message])
+    case Any
+    
+    def message: Ls[Message -> Opt[Loc]] = this match
+      case Expect.Module(msg) => msg.toList.map(_ -> N)
+      case Expect.NonModule(msg) => msg.toList.map(_ -> N)
+      case Expect.Any => Nil
+    
+    def module = this match
+      case Expect.Module(_) => true
+      case _ => false
+    
+    def nonModule = this match
+      case Expect.NonModule(_) => true
+      case _ => false
+    
+  end Expect
+  import Expect.*
+  
+  type Resolvable = Term & IResolvable
```

Comment:
Why not just
```suggestion
  type ResblTerm = Term & Resolvable
```

Reply:
My thought was that `Resolvable` and `Resolvable & Term` were equivalent because `Resolvable` is sealed and has a self-type of `Term`. It's just that Scala couldn't deduce it. So I created this funny alias, overriding the original name. 😣

Reply:
Sure. But maybe just define the trait `ResolvableImpl` and next to it add `type Resolvable = Term & ResolvableImpl`.

Local type aliases that _redefine_ already-used names are not a very good idea.